File: 0437-path-sum-iii/0437-path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
public:
    Solution()
    {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
    }
    unordered_map<long long,int> mp;
    int ans = 0;
    int pathSum(TreeNode* root, int targetSum)
    {
         mp[0]++; // when total subarray is target
         solve(root,0,targetSum);
         return ans;
    }
    void solve(TreeNode* root,long long value,int targetSum) 
    {
         if(!root) return;

         value += root->val;
         ans += mp[value - targetSum];
         mp[value]++;

         solve(root->left,value,targetSum);
         solve(root->right,value,targetSum);
         
         mp[value]--;
         value-=root->val;
    }
};

File: 0437-path-sum-iii/0437-path-sum-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number}
 */



var pathSum = function(root, target) 
{
    let map = new Map();
    let ans = 0;

    const solve = (root, sum) => {
        if(!root) return null;

        sum += root.val;

        if(sum === target) ans++;

        if(map.has(sum-target)) ans += map.get(sum-target);

        map.set(sum,(map.get(sum) || 0 )+ 1);

        solve(root.left,sum);
        solve(root.right,sum);

        map.set(sum,map.get(sum) - 1);

    }
    solve(root,0);
    return ans;
}